File: src/main/java/ExprParser/ExprTree.java
package ExprParser;

import jAADD.AADD;
import jAADD.BDD;
import jAADD.DD;

import java.util.ArrayList;
import java.util.Arrays;

/**
 * The class ExpressionTree implements an attributed syntax tree (AST).
 *
 * The base class of the expression tree is a simple value,
 * represented by result: a constant literal or a variable.
 * For variables, there is an id (name); for literals the result is its value.
 */
public class ExprTree {
    protected DD result;
    protected String id;
    protected static SymbolTable symTab;

    ExprTree()                  { result = null; id = "uninitialized"; }
    ExprTree(DD val)            { result = val;  id = ""; }
    ExprTree(String id, DD val) { setResult(val); setId(id); }
    ExprTree copy()             { return new ExprTree(getId(), getResult()); }


    public DD           getResult()                 { return result; }
    public void         setResult(DD v)             { result = v; }
    public String       getId()                     { return id; }
    public void         setId(String id)            { this.id = new String(id); }
    public static SymbolTable getSymTab()           { return symTab; }
    public static void  setSymTab(SymbolTable s)    { symTab = s; }

    AADD getAADD() throws ExprError {
        if (this.getResult() instanceof AADD) return (AADD) getResult();
        else throw new ExprError("wrong type. Expect REAL.");
    }

    BDD getBDD() throws ExprError {
        if (this.getResult() instanceof BDD) return (BDD) getResult();
        else throw new ExprError("wrong type. Expect BOOL.");
    }

    void evalThis() throws ExprError {
        if (id.equals("")) return;        // a literal.
        ExprTree v = symTab.getVar(id);
        result = v.result;
    }

    /**
     * This method computes the expression from the leavese to the root.
     * @throws ExprError
     */
    void evalExpr() throws ExprError { evalThis(); }

    /**
     * This method computes bi-directional constraint-net.
     * @throws ExprError
     */
    void propagateExpr() throws ExprError {

    }

    @Override
    public String toString() {
        if (id == null ||id.isBlank()) return ""+getResult();
        else return "("+id+": "+getResult()+")";
    }
}

File: src/main/kotlin/com/github/tuk-cps/exprParser/ExprTree.kt
package exprParser

import jAADD.AADD
import jAADD.BDD
import jAADD.DD

/**
 * The class ExpressionTree implements an attributed syntax tree (AST).
 * The base class of the expression tree is a simple value,
 * represented by result: a constant literal or a variable.
 * For variables, there is an id (name); for literals the result is its value.
 */
open class ExprTree(value: DD<*> = BDD.Infeasible,
                    id: String = "uninitialized",
                    parent: ExprTree? = null) {

    var value: DD<*>   = value   // The result computed or the value, if leaf
    var id: String      = id       // If a variable, its name.
    internal var parent = parent   // Reference to the parent node or null, if root.

    open fun copy(): ExprTree = ExprTree(value, id, parent)

    @get:Throws(ExprError::class)
    val aadd: AADD
        get() = if (value is AADD) value as AADD else throw ExprError("wrong type. Expected REAL.")

    @get:Throws(ExprError::class)
    val bdd: BDD
        get() = if (value is BDD) value as BDD else throw ExprError("wrong type. Expected BOOL.")

    @Throws(ExprError::class)
    open fun evalUp() {
        if (id == "") return  // a number literal.
        val v: ExprTree = symTab.getVar(id)
        value = v.value
    }

    open fun evalDown() {
        if (id == "") return // a number literal
        val v: ExprTree = symTab.getVar(id)
        if (this.value is AADD)  {
            this.value = (this.value as AADD).intersect(v.aadd)
        }
    }

    /** This method computes the expression from the leaves to the root. */
    @Throws(ExprError::class)
    open fun evalUpRec() = evalUp()

    open fun evalDownRec() = evalDown()

    override fun toString(): String
        = if (id.isBlank()) "" + value else "(" + id + ": " + symTab.getVar(id).value + ")"

    companion object {
        lateinit var symTab: SymbolTable
    }
}